websocket_client.py

```diff
@@ -0,0 +1,463 @@
+import argparse
+import asyncio
+import datetime
+import json
+import logging
+import os
+import signal
+import sys
+import urllib.parse
+from typing import Any, AsyncGenerator, Awaitable, Literal
+
+import aiohttp
+import numpy as np
+import pyee.asyncio
+import sounddevice
+from websockets import exceptions as ws_exceptions
+from websockets.asyncio import client as ws_client
+
+
+class LocalAudioSink:
+    """AudioSink that plays to the default audio device."""
+
+    def __init__(self, sample_rate: int = 48000, num_channels: int = 1) -> None:
+        self._sample_rate = sample_rate
+        self._num_channels = num_channels
+        self._buffer: bytearray = bytearray()
+
+        def callback(outdata: np.ndarray, frame_count, time, status):
+            output_frame_size = len(outdata) * 2
+            next_frame = self._buffer[:output_frame_size]
+            self._buffer[:] = self._buffer[output_frame_size:]
+            if len(next_frame) < output_frame_size:
+                next_frame += b"\x00" * (output_frame_size - len(next_frame))
+            outdata[:] = np.frombuffer(next_frame, dtype="int16").reshape(
+                (frame_count, num_channels)
+            )
+
+        self._stream = sounddevice.OutputStream(
+            samplerate=sample_rate,
+            channels=num_channels,
+            callback=callback,
+            device=None,
+            dtype="int16",
+            blocksize=sample_rate // 100,
+        )
+        self._stream.start()
+        if not self._stream.active:
+            raise RuntimeError("Failed to start streaming output audio")
+
+    def write(self, chunk: bytes) -> None:
+        self._buffer.extend(chunk)
+
+    def drop_buffer(self) -> None:
+        self._buffer.clear()
+
+    async def close(self) -> None:
+        if self._stream:
+            self._stream.close()
+
+
+class LocalAudioSource:
+    """AudioSource that reads from the default microphone."""
+
+    def __init__(self, sample_rate=48000, channels=1):
+        self._sample_rate = sample_rate
+        self._num_channels = channels
+
+    async def stream(self) -> AsyncGenerator[bytes, None]:
+        queue: asyncio.Queue[bytes] = asyncio.Queue()
+        loop = asyncio.get_running_loop()
+
+        def callback(indata: np.ndarray, frame_count, time, status):
+            loop.call_soon_threadsafe(queue.put_nowait, indata.tobytes())
+
+        stream = sounddevice.InputStream(
+            samplerate=self._sample_rate,
+            channels=self._num_channels,
+            callback=callback,
+            device=None,
+            dtype="int16",
+            blocksize=self._sample_rate // 100,
+        )
+        with stream:
+            if not stream.active:
+                raise RuntimeError("Failed to start streaming input audio")
+            while True:
+                yield await queue.get()
+
+
+class WebsocketVoiceSession(pyee.asyncio.AsyncIOEventEmitter):
```

Comment:
"""
    Manages a WebSocket-based voice interaction session.
    
    Handles the complete lifecycle of a voice interaction, including:
    - WebSocket connection management
    - Audio streaming (input and output)
    - State transitions (idle, listening, thinking, speaking)
    - Transcript handling
    - Client-side tool invocations

    Events emitted:
    - 'state': Current session state
    - 'output': Transcript text and completion status
    - 'error': Session errors
    - 'ended': Session completion

    Args:
        join_url (str): WebSocket connection URL for the Ultravox call. Call must have been created with `serverWebSocket` medium. (see https://docs.ultravox.ai/api-reference/calls/calls-post)
    """